In [91]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [93]:
# Import data
path = "/Users/Glebazzz/Jupiter/04.25 Instacart Basket Analysis" 
ords_prods_merge_peaks = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared data', 'ords_prods_merged_peaks.pkl'))

In [94]:
# Create one million dataframe
df = ords_prods_merge_peaks[:1000000]

In [97]:
df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   Unnamed: 0.1  order_id  user_id  order_number  orders_day_of_week  hour  \
0          1987   3139998      138            28                   6    11   
1          1989   1977647      138            30                   6    17   
2         11433    389851      709             2                   0    21   
3         12198    652770      764             1                   3    13   
4         12200   1813452      764             3                   4    17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy   Regularly busy           Most orders  
1  Mid-range product  Regularly busy   Regularly busy        Average orders  
2  Mid-range product     Busiest day     Busiest days        Average orders  
3  Mid-range product  Regularly busy  Least busy days           Most orders  
4  Mid-range product      Least busy  Least busy days        Average orders

In [99]:
df.shape

(1000000, 19)

In [101]:
# Drop additional colomn
df = df.drop(columns= ['Unnamed: 0.1'])


In [103]:
df.groupby('product_name')

In [105]:
#Performing a Single Aggregation
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.562844
2                17.518895
3                16.907152
4                17.537934
5                14.894521
6                16.658449
7                17.072954
8                14.942021
9                15.447580
10               18.681852
11               15.352278
12               14.327957
13               16.578690
14               16.666369
15               16.153766
16               17.768070
17               15.602964
18               19.674252
19               16.924344
20               16.292631
21               25.535596

In [107]:
# statistical test
df.groupby('department_id')['order_number'].mean()

department_id
1     15.562844
2     17.518895
3     16.907152
4     17.537934
5     14.894521
6     16.658449
7     17.072954
8     14.942021
9     15.447580
10    18.681852
11    15.352278
12    14.327957
13    16.578690
14    16.666369
15    16.153766
16    17.768070
17    15.602964
18    19.674252
19    16.924344
20    16.292631
21    25.535596
Name: order_number, dtype: float64

In [109]:
# Performing Multiple Aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.562844   1  99
2                17.518895   1  96
3                16.907152   1  99
4                17.537934   1  99
5                14.894521   1  99
6                16.658449   1  99
7                17.072954   1  99
8                14.942021   1  98
9                15.447580   1  99
10               18.681852   1  99
11               15.352278   1  99
12               14.327957   1  99
13               16.578690   1  99
14               16.666369   1  99
15               16.153766   1  99
16               17.768070   1  99
17               15.602964   1  99
18               19.674252   1  99
19               16.924344   1  99
20               16.292631   1  99
21               25.535596   1  98

def categorize_customers(df: pd.DataFrame) -> pd.DataFrame:
    """
    Categorize customers based on their order numbers.
    
    Args:
        df (pd.DataFrame): DataFrame containing customer order data with 'order_number' column
    
    Returns:
        pd.DataFrame: DataFrame with added 'label' column categorizing customers
    """
    # Create a copy to avoid modifying the original dataframe
    result_df = df.copy()
    
    # Define customer categories based on order numbers
    result_df.loc[result_df['order_number'] > 40, 'label'] = 'loyal customer'
    result_df.loc[(result_df['order_number'] > 10) & 
                 (result_df['order_number'] <= 40), 'label'] = 'Regular customer'
    result_df.loc[result_df['order_number'] <= 10, 'label'] = 'New customer'
    
    return result_df

# Apply the categorization
df = categorize_customers(df)

# Check the distribution of customer categories
df['label'].value_counts(dropna=False)

# Aggregating Data with transform()

In [113]:

def get_max_value(series):
    """
    Returns the maximum value in a pandas Series.
    
    Args:
        series (pd.Series): The input series to find maximum value from
        
    Returns:
        float or int: The maximum value in the series
    """
    return series.max()

# Add a column with the maximum order number for each user
ords_prods_merge_peaks['max_order'] = (
    ords_prods_merge_peaks
    .groupby(['user_id'])['order_number']
    .transform(get_max_value)
)

In [114]:
ords_prods_merge_peaks. head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    Unnamed: 0.1  order_id  user_id  order_number  orders_day_of_week  hour  \
0           1987   3139998      138            28                   6    11   
1           1989   1977647      138            30                   6    17   
2          11433    389851      709             2                   0    21   
3          12198    652770      764             1                   3    13   
4          12200   1813452      764             3                   4    17   
5          12372   1701441      777            16                   1     7   
6          13096   1871483      825             3                   2    14   
7          14603   1290456      910            12                   3    10   
8          17065    369558     1052            10                   1    20   
9          17070    589712     1052            15                   1    12   
10         22082   1798115     1379             1                   5    11   
11         23713    497193     1480            36                   3    13   
12         23715   1727826     1480            38                   4    13   
13         23716   2694698     1480            39                   2    12   
14         23937   3171753     1494             6                   3    10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy   Regularly busy           Most orders   
1   Mid-range product  Regularly busy   Regularly busy        Average orders   
2   Mid-range product     Busiest day     Busiest days        Average orders   
3   Mid-range product  Regularly busy  Least busy days           Most orders   
4

In [115]:
#running function
pd.options.display.max_rows = None

In [116]:
# checking result 
ords_prods_merge_peaks.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

# Deriving Columns with loc()

In [118]:
ords_prods_merge_peaks.loc[ords_prods_merge_peaks['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge_peaks.loc[(ords_prods_merge_peaks['max_order'] <= 40) & (ords_prods_merge_peaks['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge_peaks.loc[ords_prods_merge_peaks['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [119]:
# Get the count of each unique value in the 'Loyalty flag' column, including NaN values
ords_prods_merge_peaks['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

In [120]:
#returning specific colomns to check the result
ords_prods_merge_peaks[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

In [130]:
ords_prods_merge_peaks.groupby(['loyalty_flag']).user_id.nunique()

loyalty_flag
Loyal customer       17017
New customer        112328
Regular customer     76864
Name: user_id, dtype: int64

In [132]:
#returning specific colomns to check the result
ords_prods_merge_peaks[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

## Exercise 

In [134]:
#Performing a Single Aggregation for the whole dataset
ords_prods_merge_peaks.groupby('product_name')
ords_prods_merge_peaks.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

The first table displays the mean order_number for each department across the entire dataset, while the second table shows the mean values for a subset of the data (likely filtered by time, customer segment, or other criteria).

Key Differences:
	•	Across almost all department_ids, the mean order_number in the subset is slightly higher or lower than in the full dataset.
	•	The most notable differences are:
	•	department_id 10: from 18.68 (full) to 20.20 (subset) → a clear increase.
	•	department_id 21: from 25.54 (full) to 22.90 (subset) → a noticeable decrease.

Interpretation:
	•	These shifts indicate that customer ordering behavior varies within the subset. It might suggest:
	•	Different ordering frequency.
	•	A more/less engaged customer group.
	•	Specific departments being more/less popular in the subset.

In [138]:
# Performing Multiple Aggregations
ords_prods_merge_peaks.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457687   1  99
2                17.277920   1  99
3                17.179756   1  99
4                17.811403   1  99
5                15.213779   1  99
6                16.439806   1  99
7                17.225773   1  99
8                15.340520   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170828   1  99
12               15.887622   1  99
13               16.583304   1  99
14               16.757377   1  99
15               16.165037   1  99
16               17.663250   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

In [140]:
ords_prods_merge_peaks.shape

(32434212, 21)

# The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [142]:
# Check basic statistics of prices by loyalty_flag
ords_prods_merge_peaks.groupby('loyalty_flag')['prices'].describe()

count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10293737.0  10.388747  327.864108  1.0  4.2  7.4  11.2   
New customer       6249398.0  13.294370  597.301692  1.0  4.2  7.4  11.3   
Regular customer  15891077.0  12.496203  539.478009  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0

# The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [145]:
#Performing a Single Aggregation for prices by user_id
user_avg_prices = ords_prods_merge_peaks.groupby('user_id').agg({'prices': 'mean'})['prices']

# adding column 
ords_prods_merge_peaks['avg_value'] = ords_prods_merge_peaks['user_id'].map(user_avg_prices)

ords_prods_merge_peaks.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   Unnamed: 0.1  order_id  user_id  order_number  orders_day_of_week  ...  \
0          1987   3139998      138            28                   6  ...   
1          1989   1977647      138            30                   6  ...   
2         11433    389851      709             2                   0  ...   
3         12198    652770      764             1                   3  ...   
4         12200   1813452      764             3                   4  ...   

   add_to_cart_order  reordered  _merge    price_range_loc     busiest_day  \
0                  5          0    both  Mid-range product  Regularly busy   
1                  1          1    both  Mid-range product  Regularly busy   
2                 20          0    both  Mid-range product     Busiest day   
3                 10          0    both  Mid-range product  Regularly busy   
4                 11          1    both  Mid-range product      Least busy   

      busiest_days busiest_period_of_day max_order      loyalty_flag  \
0   Regularly busy           Most orders        32  Regular customer   
1   Regularly busy        Average orders        32  Regular customer   
2     Busiest days        Average orders         5      New customer   
3  Least busy days           Most orders         3      New customer   
4  Least busy days        Average orders         3      New customer   

   avg_value  
0   6.935811  
1   6.935811  
2   7.930208  
3   4.972414  
4   4.972414  

[5 rows x 22 columns]

In [147]:
#applying the logic to identify high and low spenders
ords_prods_merge_peaks.loc[ords_prods_merge_peaks['avg_value'] < 10, 'spenders_flag'] = 'Low spender'
ords_prods_merge_peaks.loc[ords_prods_merge_peaks['avg_value'] >= 10, 'spenders_flag'] = 'High spender'
# Get the count of each unique value in the 'Loyalty flag' column, including NaN values
ords_prods_merge_peaks['spenders_flag'].value_counts(dropna = False)

spenders_flag
Low spender     31798751
High spender      635461
Name: count, dtype: int64

In [150]:
ords_prods_merge_peaks.groupby(['spenders_flag']).user_id.nunique()

spenders_flag
High spender      5375
Low spender     200834
Name: user_id, dtype: int64

In [153]:
#returning specific colomns to check the result
ords_prods_merge_peaks[['user_id', 'loyalty_flag', 'spenders_flag']].head(50)

user_id      loyalty_flag spenders_flag
0       138  Regular customer   Low spender
1       138  Regular customer   Low spender
2       709      New customer   Low spender
3       764      New customer   Low spender
4       764      New customer   Low spender
5       777  Regular customer   Low spender
6       825      New customer   Low spender
7       910  Regular customer   Low spender
8      1052  Regular customer   Low spender
9      1052  Regular customer   Low spender
10     1379      New customer   Low spender
11     1480    Loyal customer   Low spender
12     1480    Loyal customer   Low spender
13     1480    Loyal customer   Low spender
14     1494  Regular customer   Low spender
15     1494  Regular customer   Low spender
16     1494  Regular customer   Low spender
17     1540  Regular customer   Low spender
18     1540  Regular customer   Low spender
19     1540  Regular customer   Low spender
20     1540  Regular customer   Low spender
21     1540  Regular customer   Low spender
22     1540  Regular customer   Low spender
23     1540  Regular customer   Low spender
24     1540  Regular customer   Low spender
25     1540  Regular customer   Low spender
26     1540  Regular customer   Low spender
27     1540  Regular customer   Low spender
28     1540  Regular customer   Low spender
29     1540  Regular customer   Low spender
30     1540  Regular customer   Low spender
31     1540  Regular customer   Low spender
32     1540  Regular customer   Low spender
33     1540  Regular customer   Low spender
34     1598      New customer   Low spender
35     1598      New customer   Low spender
36     1598      New customer   Low spender
37     1647      New customer   Low spender
38     1647      New customer   Low spender
39     1647      New customer   Low spender
40     1647      New customer   Low spender
41     1647      New customer   Low spender
42     2833      New customer   Low spender
43     2850  Regular customer   Low spender
44     2857  Regular customer   Low spender
45     3010  Regular customer   Low spender
46     3029      New customer   Low spender
47     3395      New customer   Low spender
48     3904      New customer   Low spender
49     4006    Loyal customer   Low spender

In [155]:
#Performing a Single Aggregation for days_since_prior_order by user_id
frequency = ords_prods_merge_peaks.groupby('user_id').agg({'days_since_prior_order': 'mean'})['days_since_prior_order']

# adding column 
ords_prods_merge_peaks['frequency'] = ords_prods_merge_peaks['user_id'].map(frequency)

ords_prods_merge_peaks.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   Unnamed: 0.1  order_id  user_id  order_number  orders_day_of_week  ...  \
0          1987   3139998      138            28                   6  ...   
1          1989   1977647      138            30                   6  ...   
2         11433    389851      709             2                   0  ...   
3         12198    652770      764             1                   3  ...   
4         12200   1813452      764             3                   4  ...   

   _merge    price_range_loc     busiest_day     busiest_days  \
0    both  Mid-range product  Regularly busy   Regularly busy   
1    both  Mid-range product  Regularly busy   Regularly busy   
2    both  Mid-range product     Busiest day     Busiest days   
3    both  Mid-range product  Regularly busy  Least busy days   
4    both  Mid-range product      Least busy  Least busy days   

  busiest_period_of_day max_order      loyalty_flag avg_value spenders_flag  \
0           Most orders        32  Regular customer  6.935811   Low spender   
1        Average orders        32  Regular customer  6.935811   Low spender   
2        Average orders         5      New customer  7.930208   Low spender   
3           Most orders         3      New customer  4.972414   Low spender   
4        Average orders         3      New customer  4.972414   Low spender   

   frequency  
0  10.400000  
1  10.400000  
2   7.070968  
3  10.948718  
4  10.948718  

[5 rows x 24 columns]

In [ ]:
# Get the count of each unique value in the 'frequency flag' column, including NaN values
ords_prods_merge_peaks['frequency'].value_counts(dropna = False)

In [159]:
# Export the updated DataFrame
ords_prods_merge_peaks.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_flags.pkl'))


In [160]:
ords_prods_merge_peaks.shape

(32434212, 24)